### PR TITLE
Spotlight 3.3.0 changes

### PR DESCRIPTION
**Pull request by sephirothkod**
# Summary

Updates this gem to be compatible with [Spotlight 3.3.0](https://github.com/projectblacklight/spotlight/tree/v3.3.0) (from 2.7.0). 

There are several tickets relevant to this PR: 
- https://github.com/harvard-lts/CURIOSity/issues/13 
- https://github.com/harvard-lts/CURIOSity/issues/17 
- https://github.com/harvard-lts/CURIOSity/issues/21

## Harvesting 

_https://github.com/harvard-lts/CURIOSity/issues/13_ 

Correctly imports resources and relies on sidecar auto creation to properly generate the sidecars.

Notable changes: 

- Convert `Spotlight::Resources::OaipmhHarvester` to `Spotlight::OaipmhHarvester` 
  - Spotlight has overhauled how they index resources, and the harvester being a resource was causing issues 
- Broke up harvesting logic into sub-methods 
- Move OAI scraping logic from `PerformHarvestsJob` to harvester model
- Move all OAI parsing logic to `OaipmhModsParser` (was `OaipmhModsItem`) 
- Create new `OaipmhUpload` resource to be an ActiveRecord mirror of the data in Solr 
- Extract all used logic from `OaipmhBuilder` and delete class 
  - Spotlight uses an ETL pipeline now instead of builders 

## JobTracking

_https://github.com/harvard-lts/CURIOSity/issues/17_ 

Notable changes: 

- Add Spotlight's built-in `JobTracking` logic to `PerformHarvestsJob` 

### Screenshots 

<details><summary>Harvesting shows up in Dashboard > Indexing activity</summary>

Curation - Dashboard dev - Harvard Curiosity 2022-07-21 at 10 50 40 AM

</details>

<details><summary>Harvesting progress tracking</summary>

Curation - Dashboard dev - Harvard Curiosity 2022-07-21 at 10 51 26 AM

</details>

## Permalinks / URNs

_https://github.com/harvard-lts/CURIOSity/issues/21_

- Replace the exhibit id & generic spotlight ID in the slug  for harvested items with the `external_id` (a column that includes both the exhibit id and the URN)
- Make the external ID uppercase so that it matches with the standard way of writing URNs (they are all uppercase now)
- When exhibits are harvested, items should all come in with an uppercase URN, regardless of if their IDs were saved as lowercased or not. 

## Miscellaneous

- Add configuration options 
  - `download_full_image` 
  - `standard_spotlight_fields` 
  - `use_iiif_images` 
- Changes to engine to allow gem migrations to work correctly in applications 
- Update specs 

# Testing Instructions 

1. In CURIOSity, `git checkout harvesting-take-3` 
2. Login as an admin 
3. Navigate to `/spotlight/new` and create a new Exhibit 
4. Click "Exhibit Dashboard" from the dropdown at the top right of the navbar 
5. Click "Items" in the left-hand sidebar 
6. Click "Add Items" 
7. Fill out the form with the following credentials 
  a. Base URL - https://api.lib.harvard.edu/oai/
  b. Set spec - demo
  c. Mapping File - upload [this file](https://github.com/harvard-lts/spotlight-oaipmh-resources/blob/job_entry/harvard_yaml_mapping_files/mods/curiosity.yml) 
8. Click "Add MODS Items" 
9. Click "Dashboard" from the left-hand sidebar 
10. Track the progress of the harvest under the "Recent item indexing activity" section 
11. Once the harvest finishes, verify that items show up in the exhibit (e.g. via the search results) 

### Interested parties: @dl-maura @phil-plencner-hl

### CODE REVIEW

**Comment by sephirothkod** [comment]
- Harvester now only runs one time per indexing. 
- When installing the gem you will have to run the migrations now to add the external_id value to resources.
- A new subclass `OaiUpload` exists specifically for items imported via this gem.
- Users will likely have to add their custom fields to the spotlight initializer.

File: app/models/spotlight/resources/oaipmh_harvester.rb
Comment (bkiahstroud) [inline app/models/spotlight/resources/oaipmh_harvester.rb:13]: I don't think this line is correct; it's still trying to index the OAI records as the harvester's own data

File: app/services/spotlight/resources/oaipmh_builder.rb
Comment (bkiahstroud) [inline app/services/spotlight/resources/oaipmh_builder.rb:22]: This is a debugging line that should be removed

Comment (sephirothkod) [reply app/services/spotlight/resources/oaipmh_builder.rb]: thanks for catching this, I've removed it.

File: app/models/spotlight/resources/oaipmh_harvester.rb
Comment (sephirothkod) [reply app/models/spotlight/resources/oaipmh_harvester.rb:13]: This line is fine, it is the to_solr method itself that is the issue. I fixed it but changed it back based on what Rob needed. I think we should hold off on changing this until we have a harvester page that we can run re-imports from. otherwise there is no way to update from the UI.

File: app/models/spotlight/resources/oaipmh_mods_item.rb
Comment (bkiahstroud) [inline app/models/spotlight/resources/oaipmh_mods_item.rb:24]: [Comment](https://github.com/harvard-lts/spotlight-oaipmh-resources/pull/7#discussion_r927002784) from @orangewolf on #7 -- 

>this is a pretty weird side effect. we should probably refactor to be more OO

**Review by orangewolf — APPROVED** [review]

**Comment by phil-plencner-hl** [comment]
Hello @bkiahstroud,

I have a few questions about how we can properly test this is working and what the next steps would be.

1. Should we be using the `harvesting-take-3` branch of the `harvard-curiosity` repository? It seems to reference the `spotlight-3.3.0-changes` branch in `spotlight-oaipmh-resources' in the Gemfile, but I wasn't sure: https://gitlab.com/notch8/harvard-curiosity/-/blob/harvesting-take-3/Gemfile#L96
2. I spun up CURIOSity using the `harvesting-take-3` branch, but I don't currently have any Exhibits created. After speaking with Vanessa Venti this morning, she has Harvesting metadata files for particular Exhibits, but didn't have any exported versions of them in CURIOSity 3.0. How do we test the importing? Do you have particular Exhibit(s) you're using and can you point to the appropriate files to use? 
3. I'm also not sure if the version I have running is really using the proper version of the Gem, but I'm admittedly not familiar with how Bundler and Rails work together. For example, if I run `bundle install` in the container I see: `Using spotlight-oaipmh-resources 0.1.2 from https://github.com/harvard-lts/spotlight-oaipmh-resources (at spotlight-3.3.0-changes@f69d870)` which appears correct. However if I run `bundle info spotlight-oaipmh-resources` it displays `/app/bundle/ruby/2.7.0/bundler/gems/spotlight-oaipmh-resources-f69d8705449c` as the path, but the files there don't seem to match what is in the `spotlight-3.3.0-changes` branch.
4. What is the path forward if we approve this PR? I'm not sure if merging this into the `main` branch will break the existing Production instances of CURIOSity / CNA since they are not using Spotlight 3.x but not 100% sure.  Looking at the repositories though it seems like they are using non-main versions of the branches: [CNA](https://github.huit.harvard.edu/LTS/CNA/blob/develop/Gemfile#L85) [CURIOISity](https://github.com/harvard-library/Curiosity/blob/harvard_main/Gemfile#L78). So we might be ok? Should we perhaps make Releases / Tags of these just to make it a little more clear what is going on in case we need to manage older instances (CNA for example) as we continue to roll this out?

I think that's enough questions for now ;) 

Thanks!

**Comment by dl-maura** [comment]
re; Number 3: @bkiahstroud we need the new Gemfile.lock. I purged mine and rebuilt and now I have the correct spotlight-oaipmh-resources-0a6f40786734

**Comment by phil-plencner-hl** [comment]
Approved as-is and merging into main knowing that there are newly created tickets to address the issues noted in my earlier comment.

**Review by phil-plencner-hl — APPROVED** [review]
Approved.

**Pull request merged**
